Add more tests of the tick tree

espendk opened the pull request:
This PR implements tests of Mangrove's tick tree to ensure that the tree is always in a consistent state.

# TODO
- [x] Once tick math supports `MIN_TICK` the tests that use it should be enabled
- [x] Once the bug described in `TickTreeUpdateOffer` is fixed, the tests that trigger it should be enabled
- [x] Consider testing what the tick tree looks like when inside `makerExecute` and `makerPosthook`.
   - `makerExecute`: We'll add a read-lock so this isn't relevant: https://github.com/mangrovedao/mangrove-core/issues/503
   - `makerPosthook`: There's a separate issue on that: https://github.com/mangrovedao/mangrove-core/issues/506
- [x] Review the `FIXMEs` added in this PR; Many of them are comments and questions that should be discussed and then the FIXME should be removed.
   - Added questions on the `FIXME`s in this PR for @adhusson  :-)
- [x] Review whether all relevant tick scenarios are included for all test contracts


# High-level test approach
For optimization reasons, the tick tree data structure is not encapsulated but read and manipulated in a low-level way in all the operations that use it (and this is interleaved with other non-tree logic). This means that we cannot unit test the data structure, but instead have to make integration tests of all the functions that manipulate the tree.

The approach taken in these integration tests is to simulate what Mangrove should be doing to its tick tree in a parallel tick tree implementation which is simpler and easier to reason about. By comparing the resulting tick trees (MGV and the test implementation) we can check that they agree on the end state; If not, we can report exactly where the discrepancy is.

In overview, the tests are structured as follows:

1. Set up Mangrove's initial state, i.e, post offers at relevant ticks (and possibly retract some of them).
2. Take a snapshot of Mangrove's tick tree, i.e, a copy in the form of a test `TickTree`
3. Perform some operation on Mangrove (e.g, add or remove an offer)
4. Perform equivalent operation(s) on the test `TickTree`
5. Compare Mangrove's tick tree to the test `TickTree`.


# Structure of the test contracts
There are multiple test contracts named `TickTree*` all located in `test/core/ticktree`.

The base contract is `TickTreeTest` and implements the test `TickTree` struct and operations on it as well as shared functions for the test contracts.

For each Mangrove operation that modifies the tick tree, there is a separate contract which tests that operation on the tick tree. For example, `newOffer*` is tested in `TickTreeNewOfferTest`.


# Test scenarios
Mangrove's tick tree is quite complex due to the different levels each being stored separately as well as some things being stored in `local`. There are there many scenarios to test, e.g, updating an offer to another tick within the same `Leaf` is different than updating it to a tick below a different level1 position.


## Tick scenarios
### Tick of interest, higher tick, and lower tick
For all the tests, there's a notion of "tick of interest", a higher tick, and a lower tick. This allows us to capture the different scenarios where ticks are in the same/different leaf, level0, level1, and level2. `TickTreeTest` contains functions for generating all relevant higher and lower ticks for a given tick of interest. Combinations of three such ticks (including the absence of higher/lower) is captured by the `TickScenario` struct.

For each test contract, these we define a `*Scenario` struct (eg `UpdateOfferScenario`) that defines what a scenario looks like and we describe how the `TickScenario` should be interpreted in these scenarios.


## Structure of test contracts
The tests contracts are therefore generally structured as follows:

1. An execute scenario function (e.g, `run_update_offer_scenario`)
   - This executes a single scenario
   - As this function may be called multiple times within a single `test_*` function, it does a `vm.snapshot()` initially and a `vm.revertTo(vmSnapshotId)` at the end. This is super fragile and if a scenario fails, you may see other, totally unrelated tests fail...
2. An execute scenarios for a specific tick (e.g, `run_update_offer_scenarios_for_tick`)
   - This runs all scenarios for a particular tick of interest
   - The reason for running multiple scenarios in one function is to avoid having to manually write many `test_*` functions that enumerate all the scenarios.
3. One or more `test_*` functions (e.g, `test_clean_offer_for_tick_0`)
   - This is the actual test function that `forge` will run.
   - The reason for having multiple `test*_` functions instead of running all scenarios in one test is that we run into limitations of `Foundry` when we do this.
   - In other words, there's a limit to how many scenarios we can run in one `test_*` function which depends on the memory used by the `run_*_scenario` function. 
   - We've tried to minimize the number of `test_*` functions, but for `updateOffer` we've had to make 4 `test_*` functions per tick we want to test :-/

adhusson committed:
doc: comment ctz

adhusson committed:
dynamic ticks

TODO:
- [ ] adapt to #437
- [ ] try to pack "OfferList info" in 1 argument everywhere for clarity
- [ ] add dynamic ticks tests
- [ ] make LogPriceLib more comfortable to use (should logPrice be a type)
-  [ ] add "what tickscales are enabled for tkn0/tkn1" to MgvReader

adhusson committed:
feat: pack out/in/scale in first argument

adhusson committed:
optim: id hashing

adhusson committed:
chore: remove DEFAULT_TICKSCALE and other cruft

adhusson committed:
chore: remove excessive OL() and other cruft

adhusson committed:
chore: Improve names

adhusson committed:
fix: copy when converting Market to OLKey

adhusson committed:
fix: set evm version to paris (later not Polgyon compatible)

adhusson committed:
fix: correct OfferWrite field name tick->logPrice

adhusson committed:
feat: events write olKeyHash, not (out,in,scale)

adhusson committed:
Update test/core/DynamicTicks.t.sol

Co-authored-by: Lasse <16150076+lnist@users.noreply.github.com>

adhusson committed:
doc: fix comment typo

adhusson committed:
doc: fix ctz spec

adhusson committed:
Update src/MgvOfferTaking.sol

Co-authored-by: Espen Højsgaard <espen@mangrove.exchange>

adhusson committed:
Update src/MgvOfferTaking.sol

Co-authored-by: Espen Højsgaard <espen@mangrove.exchange>

adhusson committed:
refactor: logPrice -> maxLogPrice for taker orders

adhusson committed:
fix: force insertionLogPrice to be a multiple of tickScale

espendk committed:
doc: Replace `snipe` with `cleanByImpersonation` in gas test README

adhusson committed:
Merge branch 'feat/ticks' into feat/dynamic-ticks-packed

lnist committed:
Re-enable gas tests

lnist committed:
adding back deleted script

lnist committed:
rename most gas-ticks to gas-logPrices

lnist committed:
make gas-measurement executable

lnist committed:
minor fixes to gas tests

adhusson committed:
chore: remove dead tests (replaced by cleaning tests)

adhusson committed:
chore: remove local dev file

adhusson committed:
fix: restore correct test

adhusson committed:
fix: match MgvLib events in IMangrove

adhusson committed:
fix: test highest value below allowed logPrice range

adhusson committed:
fix: test name

espendk committed:
wip

espendk committed:
working tests for newOffer and retractOffer

espendk committed:
cleanup + move test offer data to struct

espendk committed:
tests of update offer

espendk committed:
assert that best offer was updated correctly

espendk committed:
add snapshot of entire tick tree

espendk committed:
migrate all tests to full tick tree snapshot&comparison

espendk committed:
more remarks

espendk committed:
fix test tick tree offer insertion bug

espendk committed:
generate test scenarios for newOffer

espendk committed:
cleanup

espendk committed:
split tick tree tests into separate files

espendk committed:
remove unneeded TestMangrove

espendk committed:
comment on test scenarios for newOffer

espendk committed:
more comment on test scenarios for newOffer

espendk committed:
re-enable MAX_TICK for higher tick in scenarios

espendk committed:
simplify new offer scenarios

espendk committed:
fix test tick tree bugs

espendk committed:
cleanup

espendk committed:
implement retract tests

espendk committed:
Add tree branch to toString(Tick)

espendk committed:
cleanup

espendk committed:
Implement update offer tests

NB: The tests currently fail due to a bug in Mangrove.

espendk committed:
Add FIXMEs

espendk committed:
Improve tick tree test output

espendk committed:
comment updateOffer tick tree tests and add fuzzing

espendk committed:
fix problems with memory and inter-test interference caused by updateOffer tick tree tests

espendk committed:
Merge branch 'feat/ticks' into feat/ticks-more-tests

# Conflicts:
#	foundry.toml
#	lib/TickLib.sol
#	lib/preprocessed/ToString.post.sol
#	preprocessing/ToString.pre.sol.ts
#	src/IMangrove.sol
#	src/MgvOfferMaking.sol
#	src/MgvOfferTaking.sol
#	src/MgvRoot.sol
#	src/periphery/MgvReader.sol
#	test/lib/MangroveTest.sol
#	test/periphery/MgvReader.t.sol

espendk committed:
add tick tree clean tests

espendk committed:
add tick tree market order tests

espendk committed:
Merge branch 'feat/ticks' into feat/ticks-more-tests

espendk committed:
add tick tree test README.md

adhusson commented:
Bug fixed in #494

espendk committed:
Merge branch 'feat/ticks' into feat/ticks-more-tests

espendk committed:
enable tick tree tests of MIN_TICK

MIN_TICK can now be handled by Mangrove, so re-enabling the tests that use it.

espendk committed:
Enable tests that triggered the bug fixed in #494

espendk committed:
Convert FIXME into comment

espendk committed:
Remove deprecated FIXMEs

espendk committed:
Merge branch 'feat/ticks' into feat/ticks-more-tests

# Conflicts:
#	src/MgvRoot.sol
#	src/preprocessed/MgvLocal.post.sol

lnist commented on test/core/ticktree/README.md:
```diff
@@ -0,0 +1,56 @@
+
+# Test scenarios
+
+Mangrove's tick tree is quite complex due to the different levels each being stored separately as well as some things being stored in `local`. There are there many scenarios to test, e.g, updating an offer to another tick within the same `Leaf` is different than updating it to a tick below a different level1 position.
```
```suggestion
Mangrove's tick tree is quite complex due to the different levels each being stored separately as well as some things being stored in `local`. There are there many scenarios to test, e.g, updating an offer to another tick within the same `Leaf` is different than updating it to a tick below a different `level1` position.
```

lnist commented on test/core/ticktree/TickTreeTestAssumptions.t.sol:
```diff
@@ -0,0 +1,38 @@
+import "mgv_lib/Debug.sol";
+
+// Tests of the assumptions made in the TickTreeTests.
+// FIXME: Not sure this is needed anymore?
```
?

espendk replied:
Addressed in bda293ace9b1eb6402d3aa65425b64f821414d99

(I decided to keep this contract as self-checking documentation of the assumptions)

lnist commented on test/core/ticktree/TickTreeTestAssumptions.t.sol:
```diff
@@ -0,0 +1,38 @@
+// Tests of the assumptions made in the TickTreeTests.
+// FIXME: Not sure this is needed anymore?
+contract TickTreeTestAssumptionsTest is TickTreeTest {
+  // Cheks that the ticks used in these tests have the expected locations at various levels.
```
```suggestion
  // Checks that the ticks used in these tests have the expected locations at various levels.
```

lnist commented on test/core/ticktree/TickTreeTest.t.sol:
```diff
@@ -0,0 +1,729 @@
+// and then be compared to the actual tick tree.
+//
+// The test tick tree operations uses simpler (and less efficient) code to manipulate the tick tree, which should make
+// it clearer what is going on and easier to convince yourself that the tick tree is manipulated correctly.
```
reference readme.md here?

espendk replied:
Addressed in 917b25d6dc07e36529cff757f41114e03f2aa207

lnist commented on test/core/ticktree/TickTreeTest.t.sol:
```diff
@@ -0,0 +1,729 @@
+  function assertMgvTickTreeIsConsistent() internal {
+    Field level2 = mgv.level2(olKey);
+    for (uint level2Pos = 0; level2Pos <= MAX_LEVEL2_POSITION; ++level2Pos) {
+      if (!isBitSet(level2, level2Pos)) {
```
does it also assert that it should be set if it is not?

espendk replied:
Good point. I added a check that the level below is empty if the bit isn't set: 7e8a3a31181a4d6f5bdbb527feec4d40fbd8ef9b

I also implemented a check that the entire subtree was empty, but that used too much gas, so I compromised on one level below.

lnist commented on test/core/ticktree/TickTreeTest.t.sol:
```diff
@@ -0,0 +1,729 @@
+      }
+
+      int level1Index = level1IndexFromLevel2Pos(level2Pos);
+      Field level1 = mgv.level1(olKey, level1Index);
```
If these loops OOM then maybe putting each level in its own function will help it release memory?

espendk replied:
Worth trying if we run into the issue in the future. For now, the checks succeed so I think we should park it here :-)

lnist commented on test/core/ticktree/TickTreeMarketOrder.t.sol:
```diff
@@ -0,0 +1,243 @@
+  }
+
+  // (list size, offers to take)
+  uint[2][] tickListScenarios = [[0, 0], [1, 1], [2, 1], [2, 0]];
```
so we never take 2 offers?

espendk replied:
Indeed; maybe we should?

Maybe a single test that ensures we can take two offers would be sufficient to test the tick list updates in market orders?

espendk replied:
Tick list scenarios have been changed such that fully taken lists have length 2: 150c33c1bb456ff1f982118d155bdf2c65917f4a

lnist commented on test/core/ticktree/TickTreeMarketOrder.t.sol:
```diff
@@ -0,0 +1,243 @@
+    TickTree storage tickTree = snapshotTickTree();
+
+    // 4. Run the market order
+    mgv.marketOrderByLogPrice(olKey, MAX_LOG_PRICE, fillVolume, true);
```
Consider (but maybe separate PR) whether it is worth it to snapshot the resulting ticktree inside posthook of the last taken offer, so that we verify tree is right already when posthooks are executed. Not high priority.

espendk replied:
Good idea.  I've created an issue on it: https://github.com/mangrovedao/mangrove-core/issues/506

lnist reviewed (commented):
Very thorough tests - good decision to make a simple implementation to compare against the other-code-interleaved-version.

espendk committed:
Update test/core/ticktree/README.md

Co-authored-by: Lasse <16150076+lnist@users.noreply.github.com>

espendk committed:
Update test/core/ticktree/TickTreeTestAssumptions.t.sol

Co-authored-by: Lasse <16150076+lnist@users.noreply.github.com>

espendk committed:
Consistent use of representative ticks in tick tree tests

espendk committed:
Change market order tick tree tests to take more than 1 offer in a list

espendk committed:
Add reference to README.md

espendk committed:
Check level below is empty if bit is unset in MGV consistency check

Costs more gas, so must be explicitly commented in when needed.
A complete check of all levels below an empty bit is too expensive, so we only check the level immediately below.

espendk committed:
Change FIXME to regular comment

espendk commented on lib/Constants.sol:
```diff
@@ -33,6 +33,18 @@ int constant NUM_LEVEL0 = NUM_LEVEL1 * LEVEL1_SIZE;
 int constant NUM_LEAFS = NUM_LEVEL0 * LEVEL0_SIZE;
 int constant NUM_TICKS = NUM_LEAFS * LEAF_SIZE;
 
+// FIXME: Should these be here or placed somewhere else? Besides being useful in tests, they serve as documentation for the datastructure
```
@adhusson Do you have an opinion on where these constants should be?

adhusson replied:
If they're not used by production code I don't think they should be there. Documentation can help fill in the gaps.

espendk replied:
Addressed in 9f6a98e5c4f29483f8edb7959876ee640497e7ab

espendk commented on lib/TickLib.sol:
```diff
@@ -209,7 +209,9 @@ library TickLib {
   // note with int24 tick we only use 64 bits of level2 (64*256*256*4 is 2**24)
   // the goal is that have the bit positions in {} used:
-
+// FIXME: This is a non-standard way of writing bit positions, normally they're numbered right-to-left
+// FIXME: Also, the highest position in a field is 63 and is used for the highest tick in the level.
+//        This feels like it's conflicting with the stmt in the Excalidraw that says "ticks go right (cheap) to left (expensive)"
```
Maybe we can leave this comment here for when we update all comments?
@adhusson wdyt?

adhusson replied:
Yes the diagram is in reverse, it should go the other way, so we can leave the comment there and reverse it later!

espendk commented:
Thanks for the review @lnist ! I addressed your comments/suggestions :-)

adhusson commented on lib/preprocessed/ToString.post.sol:
```diff
@@ -56,6 +56,10 @@ function toString(LocalUnpacked memory __unpacked) pure returns (string memory)
 }
 
+function tickBranchToString(Tick tick) pure returns (string memory) {
+  return string.concat(vm.toString(tick.posInLevel2()), "->", vm.toString(tick.posInLevel1()), "->", vm.toString(tick.posInLevel0()), "->", vm.toString(tick.posInLeaf()));
```
I would go even further and give the indices of each level!

espendk replied:
Addressed in 69f8b815afb701420ae5537f9639656fca2a5bf7

adhusson commented on test/core/ticktree/TickTreeTest.t.sol:
```diff
@@ -0,0 +1,765 @@
+  struct Offer {
+    MgvStructs.OfferPacked offer;
+    MgvStructs.OfferDetailPacked detail;
+  }
```
Why not reuse MgvRoot.OfferData? (maybe not important)

espendk replied:
Good point. Originally the `Offer` struct was different from `MgvRoot.OfferData` but ended up being identical w/o me realizing 😅

Addressed in 5fd47a7ba1d39d980b20ebf27a2d70752dfe895c

adhusson commented on test/core/ticktree/TickTreeTest.t.sol:
```diff
@@ -0,0 +1,765 @@
+    MgvStructs.OfferDetailPacked detail;
+  }
+
+  struct TickTree {
```
I think this and everything below might deserve to live in a `test/lib/*.sol` file since it's pretty crucial tooling that is way too buried in a .t.sol. Or at least in a .sol in the `test/core/ticktree` dir.

espendk replied:
Ok, I'll move it 👍

espendk replied:
Addressed in 8bdd91f4a084d87f0110bf15b57966515c88269e

adhusson commented on test/core/ticktree/TickTreeUpdateOffer.t.sol:
```diff
@@ -0,0 +1,347 @@
+  //     }),
+  //     true
+  //   );
+  // }
```
This can be uncommented now

espendk replied:
That specific scenario is already part of the `tick=0` tests :-) I've kept the commented out single-scenario tests as an easy way to run a single scenario.

But now that I think about it, it would ease maintenance if those tests are commented in. I'll do that!

espendk replied:
Addressed in 794c06c0a341dd0d5757f7989e33eb67187d56bb

adhusson commented on test/core/ticktree/TickTreeUpdateOffer.t.sol:
```diff
@@ -0,0 +1,347 @@
+    }
+
+    // 1. Capture VM state before scenario so we can restore it after
+    uint vmSnapshotId = vm.snapshot();
```
This is unreliable, right? Alternatives:
- Actually create a ton of tests (also alleviates the memory leak problems in loops)
- Still loop, but create a new Mangrove every time, so no need to reset a state

espendk replied:
I think it is only unreliable when tests run out of gas/memory. When the tests work, they do so reliably.

I think we should keep it as-is for now and move on. If we need to update/change the tick tree tests at some point, we can look into improving this. Ok?

adhusson commented on foundry.toml:
```diff
@@ -12,6 +12,7 @@ fs_permissions = [{ access = "read-write", path = "./addresses/"}, { access = "r
 solc_version="0.8.21"
 ffi=true
 optimizer=false
+memory_limit = 100000000 # 33554432 is default limit
```
Depending on how hard it is, adding manual memory management before merging could be nice. Otherwise this is a ticking time bomb, + it probably slows tests.

espendk replied:
I agree that this is not ideal. But as mentioned above, I think we should wrap up this test suite and park improvements for later, as we don't have much time and much to do.

adhusson reviewed (changes requested):
Super important part of the test infrastructure, already caught a bug, just a few remarks bug apart from that let's go 🚀

espendk committed:
move constants for tick tree tests to TickTreeTest.t.sol

espendk committed:
uncomment single scenario tick tree tests for easier maintenance

espendk committed:
use MgvRoot.OfferData instead of custom Offer struct

espendk committed:
add indexes when printing tree branches

espendk committed:
Move test tick tree to separate contract

adhusson reviewed (approved):
✅✓✔☑
